xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -3498,11 +3509,154 @@ public void readObjectsFromForm(EditForm eform, XWikiContext context) throws XWi
+     * given value. If the object doesn't exist in the document, it is created then the property <propertyname> is
+     * initialized with the given value. An object is only created if the given '<number>' is 'one-more' than the
+     * existing number of objects. For example, if the document already has 2 objects of type 'Space.Class', then it
+     * will create a new object only with 'Space.Class_2_prop=something'. Every other parameter like
+     * 'Space.Class_42_prop=foobar' for example, will be ignore.
+     * 
+     * @param eform is form information that contains all the query parameters
+     * @param context
+     * @throws XWikiException
+     * @since 7.0RC1
+     */
+    public void readObjectsFromFormUpdateOrCreate(EditForm eform, XWikiContext context) throws XWikiException
+    {
+        Map<DocumentReference, SortedMap<Integer, Map<String, String[]>>> fromRequest =
+            parseRequestUpdateOrCreate(eform.getRequest(), context);
+        for (Entry<DocumentReference, SortedMap<Integer, Map<String, String[]>>> requestClassEntries : fromRequest
+            .entrySet()) {
+            DocumentReference requestClassReference = requestClassEntries.getKey();
+            SortedMap<Integer, Map<String, String[]>> requestObjectMap = requestClassEntries.getValue();
+            List<BaseObject> newObjects = new ArrayList<>();
+            for (Entry<Integer, Map<String, String[]>> requestObjectEntry : requestObjectMap.entrySet()) {
+                Integer requestObjectNumber = requestObjectEntry.getKey();
+                Map<String, String[]> requestObjectPropertyMap = requestObjectEntry.getValue();
+                BaseObject oldObject = getXObject(requestClassReference, requestObjectNumber);
+                if (oldObject == null) {
+                    // Create the object only if it has been numbered one more than the number of existing objects
+                    int objectNumberInPage = getXObjectSize(requestClassReference);
+                    if (requestObjectPropertyMap != null) {
+                        while(newObjects.size() < objectNumberInPage) {
+                            newObjects.add(getXObject(requestClassReference, newObjects.size()));
+                        }
+                        oldObject = newXObject(requestClassReference, context);
+                    } else {
+                        break;
+                    }
+                }
+                BaseClass baseClass = oldObject.getXClass(context);
+                BaseObject newObject = (BaseObject) baseClass.fromMap(requestObjectPropertyMap, oldObject);
+                newObject.setNumber(oldObject.getNumber());
+                newObject.setGuid(oldObject.getGuid());
+                newObject.setOwnerDocument(this);
+                while(newObjects.size() < requestObjectNumber) {
+                    newObjects.add(null);
+                }
```

Comment:
IMO, it would be nice to not expose this implementation detail (i.e. that unused object numbers are filled with `null`s inside the list of objects).

When I mentioned this to you in my previous commit, I only meant it as an explanation to what is going on in the background and reuse that existing method (instead of reproducing its behavior).

What it boils down to is the fact that, instead of managing the list of `newObjects`, with adding existing objects, adding new objects and filling gaps with `null`s and, in the end, replacing the entire list of objects for `requestClassReference` with `newObjects`, you could have avoided managing all these implementation details and you could have just used the API method I pointed out.

More precisely, at the end of this for loop, you would just say `setXObject(newObject.getNumber(), newObject)`. 
This way you properly handle both existing objects and new objects, since existing objects will be replaced with the updated version at the existing index and new objects will be set at the new index, while "holes" will be filled with `null`s by the current implementation of the API (`setXObject` method).

The result in code is that you can remove the while loop above (that adds `null`s), remove the `getXObjects().put(requestClassReference, newObjects);` below and also remove the while loop at line 3628 above (which adds existing objects to your re-managed list) and anything related to the handling of the `newObjects` list, since you don`t need to care about all those details any more.

I am going to accept this PR as it is, since you have fixed the main problem (which was the existing objects bug) and it is in a good shape now, but it would be really nice if you found the time to remove the useless exposure of technical details that I detailed above, since otherwise it makes the code more resistant to future change in the API's implementation.